File: clojure/70.clj
; 8319823
; 29.82user 0.95system 0:22.83elapsed 134%CPU (0avgtext+0avgdata 0maxresident)k
; 0inputs+16outputs (0major+187686minor)pagefaults 0swaps

(load-file "lib/core.clj")
(use
  '[clojure.math.combinatorics]
  '[clj-euler.core :only (isqrt load-primes)])

(def primes
  (load-primes "../data/primes.txt"))

(def limit (int 1e7))

; Search around the radius of (sqrt 1e7)
(def ps
  (take-while #(<= % (* 2 (inc (isqrt limit)))) primes))

; Create pairs for semi-primes
(def pairs
  (combinations ps 2))

(defn permutation?
  "Returns true if a's digits are a permutation of b's."
  [a b]
  (let [digits #(sort (into [] (str %)))]
    (= (digits a) (digits b))))

(defn phi
  "Calculates the totient of a semi-prime created from p and q."
  [p q]
  (* (dec p) (dec q)))

(defn f
  "Creates a map entry of n/phi(n) -> [p q]."
  [pair]
  (let [p (first pair)
        q (second pair)
        n (* p q)]
    [(/ n (phi p q)) [p q]]))

(defn g
  "Returns true if the totient of the semi-prime generated from [p q] is a
  permutation of that semi-prime."
  [pair]
  (let [p (first pair)
        q (second pair)
        n (* p q)]
    (permutation? n (phi p q))))

; Create a map of n/phi(n) -> [p q], filtering out semi-primes who are
; larger than 1e7
(def h
  (filter #(<= (apply * (second %)) 10000000)
          (apply merge (cons {} (map f pairs)))))

(let [entry (first (drop-while #(not (g (second %))) (sort h)))
      pair (second entry)
      p (first pair)
      q (second pair)]
  (println (* p q)))

File: src/70.ts
import primes from './core/primes';

// Computes the value of n for which φ((n) is a permutation of n and the ratio n/φ(n) produces a minimum.
//
// See {@link https://projecteuler.net/problem=70}
export default function compute(limit: number) {
  const pairs = getSemiPrimePairs(getLimit(limit))
    .map(pair => transformPair(pair))
    .filter(pair => isWithinLimit(pair, limit))
    .filter(pair => isPermutation(pair));

  let minRatio = limit;
  let result = 0;

  for (const [phi, n] of pairs) {
    const ratio = n / phi;
    if (ratio < minRatio) {
      minRatio = ratio;
      result = n;
    }
  }

  return result;
}

// This is a special totient function that optimizes the computation of totient for a semi-prime whose factors are p
// and q.
//
// For a prime number, the totient is simply (sub1 p).  Because the totient is a multiplicative function,
// totient(p * q) = totient(p) * totient(q).  Therefore, the totient of a semi-prime is simply (p - 1) * (q - 1).
function totient(p: number, q: number): number {
  return (p - 1) * (q - 1);
}

// Checks if two positive integers' digits are permutations of each other.
function isPermutation(pair: [number, number]): boolean {
  const [m, n] = pair;
  return m.toString().split('').sort().join('') === n.toString().split('').sort().join('');
}

function getLimit(limit: number): number {
  return 2 * (Math.floor(Math.sqrt(limit)) + 1);
}

function isWithinLimit(pair: [number, number], limit: number): boolean {
  return pair[1] < limit;
}

function getSemiPrimePairs(limit: number): [number, number][] {
  const ps = primes().filter(p => p < limit);
  const pairs: [number, number][] = [];

  for (let i = 0; i < ps.length; i++) {
    for (let j = i; j < ps.length; j++) {
      const p = ps[i];
      const q = ps[j];
      pairs.push([p, q]);
    }
  }

  return pairs;
}

// Creates totient to semi-prime pairs.
function transformPair(factors: [number, number]): [number, number] {
  const [p, q] = factors;
  const n = p * q;
  return [totient(p, q), n];
}
